site-src/geps/gep-1323.md

```diff
@@ -0,0 +1,274 @@
+          remove:
+          - foo
+      weight: 75
+```
+
+* Contour’s `HTTPProxy`:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: basic
+spec:
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes:
+    - conditions:
+      - prefix: /
+      services:
+        - name: s1
+          port: 80
+      responseHeadersPolicy:
+        set:
+          name: test
+          value: true
+```
+
+* Ingress NGINX:
+
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: nginx-headers
+  annotations:
+    nginx.ingress.kubernetes.io/configuration-snippet: |
+      add_header ingress nginx;
+spec:
+  ingressClassName: nginx
+  rules:
+  - host: custom.configuration.com
+    http:
+      paths:
+      - path: /
+        pathType: Prefix
+        backend:
+          service:
+            name: http-svc
+            port:
+              number: 80
+```
+
+## Policy Attachment
+
+This feature can be further extended via [Policy Attachment](../references/policy-attachment.md). Implementations can introduce a new Policy, `ResponseHeaderPolicy`, which can be used to ensure that certain headers are present/absent in a response, regardless of the `responseHeaderModifier` field in the `HTTPRoute` associated with that response.
+
+```golang
+type ResponseHeaderPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of ResponseHeaderPolicy.
+    Spec ResponseHeaderPolicySpec `json:"spec"`
+
+    // Status defines the current state of ResponseHeaderPolicy.
+    Status ResponseHeaderPolicyStatus `json:"status,omitempty"`
+}
+
+type ResponseHeaderPolicySpec struct {
+    // TargetRef identifies an API object to apply policy to.
+    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`
+
+    // Override defines policy configuration that should override policy
+    // configuration attached below the targeted resource in the hierarchy.
+    // +optional
+    Override *HTTPHeaderModifier `json:"override,omitempty"`
+
+    // Default defines default policy configuration for the targeted resource.
+    // +optional
+    Default *HTTPHeaderModifier `json:"default,omitempty"`
+}
+
+type ResponseHeaderPolicyStatus struct {
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+`.spec.default` provides a way to modify headers of responses being returned from the resource specified in `.spec.targetRef`. This resource can be a part of the Gateway API like `Gateway` or a core Kubernetes resource like `Namespace`. Using a `Namespace` as the `.spec.targetRef` would result in any response being returned from any resource in that namespace having its headers modified according to the policy.
+`.spec.override` can be used to modify the headers of responses being returned from a resource that is already covered by `.spec.default` of a `ResponseHeaderPolicy`. Let's walk through an example:
+
```

Comment:
This is a good example. If the community decides to keep these details within this GEP, then I'll offer a suggestion for a different way to implement a global policy with the possibility of override.  The policy is set at the Gateway to apply to all *Routes globally as you described, but then each *Route could override the global default if needed, using the `ResponseHeaderModifier` mechanism you described earlier.  The policy would then need to include some way to describe whether it is optional or not, i.e. whether it can be overridden or not.  To me, this approach is clearer, but I still think it could be discussed in a different GEP.

Reply:
If we allow a Route to override the `override` set by a Gateway using its `ResponseHeaderModifier`, how would admins set headers alongside the headers specified  in the Route's `ResponseHeaderModifier`?
About, moving the Policy Attachment section to its own GEP, I'm neither for or against it, so some consensus from the maintainers on this would be helpful.
wdyt? @shaneutt @robscott @youngnick

Reply:
I like this proposed approach and _think_ it's sufficiently clear. [Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/) already has the structure for separating defaults and overrides - if a header was added or modified by an `Default` config a filter on the Route object should be able to further modify or remove it, but if added or modified by an `Override` config, a filter on the Route should _not_ be able to modify or remove it.

The main reason this may warrant consideration as a separate GEP is that it contradicts the current warning in https://gateway-api.sigs.k8s.io/references/policy-attachment/#interaction-with-custom-route-filters, although that does explicitly state:
> Note that this guidance may change in the future as we gain a better understanding of how extension mechanisms of the Gateway API can interoperate.

Reply:
@mikemorris having gone through the text, I understand that we disallow overlaps between Policy and _custom_ Route filters. Since this proposed filter is not custom I feel like the current proposal is more or less okay.
I'll update it with your suggestion about Routes being able to modify `Default` config. @candita I now realize that I misread your initial comment 🤦 so yes, a Route should be able to modify the gloabal default